Refactor on_mob_death, death implants, implant permission

coiax opened the pull request:
There is a proc on `/obj/item` called `on_mob_death` called on all
items in the contents of a mob on that mob's death. It is currently used
for explosive implant detonation, and the deactivation of the
Peaceborg's projectile dampener.

Instead of using this old proc, both of them now instad use the
COMSIG_LIVING_DEATH signal, already emitted when their owner dies.

The activation of an explosive implant will now occur after the rest of
the death code has run, since it activates with an async applied
function, since some other implants may still want the mob's body
intact, and you shouldn't use `sleep()` (which it does in the "slow
explosion mode") in signal handlers.

In addition, the "can_be_implanted" proc for /mob/living (and overriden
for silicons, slimes and simple animals) has been folded into the
`/obj/item/implant/proc/can_be_implanted_to` proc. Some future implants
may want to be more permissive than the current permissions, but that
isn't possible when checking both procs.

coiax committed:
Refactor on_mob_death and death implants

There is a proc on `/obj/item` called `on_mob_death` called on all
items in the contents of a mob on that mob's death. It is currently used
for explosive implant detonation, and the deactivation of the
Peaceborg's projectile dampener.

Instead of using this old proc, both of them now instad use the
COMSIG_LIVING_DEATH signal, already emitted when their owner dies.

The activation of an explosive implant will now occur after the rest of
the death code has run, since it activates with an async applied
function, since some other implants may still want the mob's body
intact, and you shouldn't use `sleep()` (which it does in the "slow
explosion mode") in signal handlers.

In addition, the "can_be_implanted" proc for /mob/living (and overriden
for silicons, slimes and simple animals) has been folded into the
`/obj/item/implant/proc/can_be_implanted_to` proc. Some future implants
may want to be more permissive than the current permissions, but that
isn't possible when checking both procs.

Jared-Fogle commented on code/game/objects/items/implants/implant.dm:
```diff
@@ -14,28 +14,28 @@
+		return TRUE
 
+	if(isanimal(target))
+		var/mob/living/simple_animal/A = target
+		// Robots and most non-organics aren't healable.
+		return A.healable
```
```suggestion
		var/mob/living/simple_animal/animal = target
		// Robots and most non-organics aren't healable.
		return animal.healable
```

Jared-Fogle commented on code/game/objects/items/implants/implant.dm:
```diff
@@ -14,28 +14,28 @@
-/mob/living/silicon/can_be_implanted()
-	return FALSE
+/obj/item/implant/proc/can_be_implanted_in(mob/living/target)
+	// Originally mob types could specify their own "implant permission" code, but it wasn't used
+	// anywhere, and neither was this proc. Folding the two together means you can have an
+	//implant that always works without too much jumping through hoops.
```
I don't think we need to document what code *used* to be, we have a lot of this and it loses its relevancy really fast.

coiax committed:
Make requested changes; comments, variable names

- Renamed a simple animal variable, to make it clearer.
- Removed a "historical note" comment.
